crates/bevy_utils/src/anystack.rs

```diff
@@ -0,0 +1,194 @@
+struct AnyStackMeta<T> {
+    offset: usize,
+    func: unsafe fn(value: *mut u8, user_data: &mut T),
+}
+
+pub struct AnyStack<T> {
+    bytes: Vec<u8>,
+    metas: Vec<AnyStackMeta<T>>,
+}
+
+impl<T> Default for AnyStack<T> {
+    fn default() -> Self {
+        Self {
+            bytes: vec![],
+            metas: vec![],
+        }
+    }
+}
+
+// SAFE: All values pushed onto the stack are required to be [`Send`]
+unsafe impl<T> Send for AnyStack<T> {}
+
+// SAFE: All values pushed onto the stack are required to be [`Sync`]
+unsafe impl<T> Sync for AnyStack<T> {}
+
+impl<T> AnyStack<T> {
+    ////Constructs a new, empty `AnyStack<T>`.
+    /// The stack will not allocate until elements are pushed onto it.
+    #[inline]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Returns the number of elements in the [`AnyStack`], also referred to as its ‘length’.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.metas.len()
+    }
+
+    /// Returns true if the [`AnyStack`] contains no elements.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Push a new `value` onto the stack.
+    ///
+    /// # Safety
+    ///
+    /// `func` must safely handle the provided `value` bytes.
+    /// _NOTE_: the bytes provided to the given function may **NOT** be aligned. If you wish
+    /// to read the value, consider using [`std::ptr::read_unalinged`].
+    ///
+    /// [`AnyStack`] does not drop the contained member.
+    /// The user should manually call [`AnyStack::consume`] and in the implementation
+    /// of the provided `func` the element should be dropped.
+    #[inline]
+    pub unsafe fn push<U>(&mut self, value: U, func: unsafe fn(value: *mut u8, user_data: &mut T))
+    where
+        U: Send + Sync,
```

Comment:
I don't *think* the `Sync` bound is strictly required here, even if `AnyStack` is `Sync`, since we only give out 'unique' references to the value, and only when you have a unique reference to ourselves.